[#455] use lan id and uid_ems to uniquely identify network

priley86 opened the pull request:
fixes #455 

* Updated the network grouping logic to now group networks by {cluster id}\{uid_ems}.
* Update the Network mapping display to show provider name in the tree view of mapped networks (since now we can have source networks with the same name but different providers mapped to the same target)
* Updates the Infrastructure Mapping List to retrieve clusters by ID coming from `/api/clusters` instead of using `transformation_mapping_id` (this allows those mappings to look across clusters)
* Updated network step snapshots

Before:
<img width="899" alt="screen shot 2018-07-06 at 12 37 49 pm" src="https://user-images.githubusercontent.com/4237045/42390442-e5b365fe-8119-11e8-9eb5-e47363802817.png">


After:
<img width="904" alt="screen shot 2018-07-06 at 11 53 25 am" src="https://user-images.githubusercontent.com/4237045/42390154-0e13d0b6-8119-11e8-96f8-ad656e93e7f2.png">


<img width="1172" alt="screen shot 2018-07-06 at 11 53 56 am" src="https://user-images.githubusercontent.com/4237045/42389897-2415a7a0-8118-11e8-97be-d48e5b44c8c6.png">

AllenBW commented on app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/components/NetworksStepForm/NetworksStepForm.js:
```diff
@@ -304,12 +305,14 @@ class NetworksStepForm extends React.Component {
                 <DualPaneMapperListItem
                   item={sourceNetwork}
-                  text={`${sourceNetwork.providerName} \\ ${sourceNetwork.name}`}
+                  text={`${sourceNetwork.providerName} \\ ${sourceNetwork.name} \\ ${__('Switch')} ${
```
Does this need the context afforded by wrapping the whole text in `sprintf`?

AllenBW commented on app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/components/NetworksStepForm/helpers.js:
```diff
@@ -42,21 +43,21 @@ export const targetNetworkWithTreeViewAttrs = targetNetwork => ({
 export const sourceNetworkWithTreeViewAttrs = (sourceNetwork, selectedCluster) => ({
   ...sourceNetwork,
-  text: sourceNetwork.name,
+  text: `${sourceNetwork.name} \\ ${__('Switch')} ${sourceNetwork.switch_id}`,
```
Same as above,  Does this need the context afforded by wrapping the whole text in sprintf?

priley86 commented:
Yes - @michaelkro agree this is a valid question. I don’t know if the UI should make this restriction but we can certainly ask.

AparnaKarve commented:
@michaelkro Yes, agree with you.

@priley86 Apologies for the late response.
What we have today (not displaying the switch id's) is what we need to retain - because that gives us the ability to automatically map multiple switches from source to target via our dedupe process, and that is the desired behavior.

So if we take `VM Network` as an example. It has the following 2 instances -

- `vcenter-lab \ VM Network`  
- `vcenter-cfme \ VM Network`

Mike and I briefly discussed this issue and we thought that the network id (LAN id) can give us the required uniqueness and serve the purpose of differentiating the 2 Networks.
I guess Cluster id would also work if that seems more appropriate.

priley86 commented:
@AparnaKarve @michaelkro i've updated the mapping logic to no longer use Lan ID and instead use Cluster ID. This should ensure the source network switches are all mapped to the same target, and enable us to map source networks with the same name.

updated PR comments and screenshots above... should be ready for testing.

AparnaKarve commented on app/javascript/react/screens/App/common/networkKey.js:
```diff
@@ -0,0 +1,4 @@
+/**
+ * returns a unique network based on current UI model
+ */
+export const networkKey = network => `${network.clusterId}\\${network.uid_ems}`;
```
I really love this idea -- `networkKey`

priley86 replied:
@AparnaKarve i'm out of fresh ideas after this one unfortunately... borderline rocket surgery here 🤣

AparnaKarve replied:
The Network id (LAN id) also seems to work here just fine! 
```js
`${network.id}\\${network.uid_ems}`
```

Really good thinking here -- isolating `networkKey` in a separate file !!

AparnaKarve commented:
Thanks everyone for your feedback!
Merging...

simaishi commented:
@priley86 There are lots of conflicts backporting to Gaprindashvili... Please create a separeate PR for G-branch.

```
$ git status
On branch gaprindashvili
Your branch is ahead of 'origin/gaprindashvili' by 3 commits.
  (use "git push" to publish your local commits)

You are currently cherry-picking commit 2102048.
  (fix conflicts and run "git cherry-pick --continue")
  (use "git cherry-pick --abort" to cancel the cherry-pick operation)

Changes to be committed:

	modified:   app/javascript/components/index.js
	modified:   app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/MappingWizardNetworksStepActions.js
	modified:   app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/__tests__/MappingWizardNetworksStepSelectors.test.js
	modified:   app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/__tests__/__snapshots__/MappingWizardDatastoresStepReducer.test.js.snap
	modified:   app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/__tests__/__snapshots__/MappingWizardNetworksStepSelectors.test.js.snap
	modified:   app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/components/NetworksStepForm/__tests__/__snapshots__/helpers.test.js.snap
	modified:   app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/mappingWizardNetworksStep.fixtures.js
	new file:   app/javascript/react/screens/App/common/networkKey.js

Unmerged paths:
  (use "git add <file>..." to mark resolution)

	both modified:   app/javascript/react/screens/App/Overview/components/InfrastructureMappingsList/helpers.js
	both modified:   app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/MappingWizardNetworksStepSelectors.js
	both modified:   app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/components/NetworksStepForm/NetworksStepForm.js
	both modified:   app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/components/NetworksStepForm/__tests__/NetworksStepForm.test.js
	both modified:   app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/components/NetworksStepForm/__tests__/helpers.test.js
	both modified:   app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/components/NetworksStepForm/helpers.js
	both modified:   app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep/index.js
```

priley86 commented:
ok - will try to rebase this for G now...

priley86 commented:
@simaishi - it looks like this PR must come first: https://github.com/ManageIQ/miq_v2v_ui_plugin/pull/395

Added `gaprindashville/yes` to that PR. It seems to rebase fine - will attempt this one again after...

simaishi commented:
Gaprindashvili backport details:

```
$ git log -1
commit 17e475e6cdf072391b8c2b8bcd38985449651ec0
Author: Aparna Karve <aparna.karve@gmail.com>
Date:   Fri Jul 6 10:45:45 2018 -0700

    Merge pull request #470 from priley86/network-identification
    
    [#455] use lan id and uid_ems to uniquely identify network
    (cherry picked from commit 2102048f251130e66c2c79d6ae64cce80dd0a6f6)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1599864
```
